### ISSUE TITLE
Usage of root while using core modules

### ISSUE THREAD

**Issue by god-of-hellfire** [ansible/ansible#22784]
##### ISSUE TYPE

 - Bug Report

##### COMPONENT NAME
- all core modules

##### ANSIBLE VERSION

```
latest
```

##### OS / ENVIRONMENT
- AIX 7.1

##### SUMMARY
Core modules like copy need root access on the remote machine to change the owner and permissions of the copied files and directories. Even if the become_user has the privileges to create such files and directories. This behavior leads to nasty shell/ command workarounds that moves the core-module-copied files to the right place in the name of the become_user. We use this workaround because of we think that granting root access to a deployment user exposes a security vulnerability in enterprise scenarios. Our proposal is to flag core modules, if root should be used or not. In case of the root-flag is false the files should be copied right into the target location as become_user with default permissions.

```yaml

```

```

```

**Comment by f-koehler** [comment]
I ran into this problem today when using the `shell` module (ansible version 2.2.1.0). It seems that `become_user` alone is not sufficient and you also have to set `become: yes`.
If I use
```yaml
- shell: echo "$USER"
  become_user: ansible
```
I see (when runing with `-vvv`) that the user `root` is used. When using
```yaml
- shell: echo "$USER"
  become_user: ansible
  become: yes
```
the script will be run with the correct user `ansible`.

Still, I dislike this behaviour. I expected that shell would run scripts as the `remote_user` from its home directory. I agree that it is problematic to run scripts by `root` by default. The only reason I detected this issue, is that I tried to execute a program that has a safety mechanism to prevent from being executed as `root`.

**Comment by god-of-hellfire** [comment]
We assume that ansible runs these steps or simliar ones to copy a local file from local to remote:
1. scp the file from local to remote temp dir with ${USER}
2. cp remote file from temp dir to destination folder as ${USER} by ssh
3. chown and chmod remote file at destination as root by ssh

`user@local$ scp file.txt user@remote.host:/tmp/.`
`user@remote$ cp /tmp/file.txt /destination/folder`
`user@remote$ sudo chown user user_group /destination/folder/file.txt`
`user@remote$ sudo chmod 744 user user_group `/destination/folder/file.txt`

If the ${USER} is member of the same group as the become_user and the become_user has access to the temp dir the steps 2 and 3 can be substitued by a remote copy of the file in the name of the become_user. This will automatically set the permissions and the group membership. So root is not necessary anymore.

**Comment by bcoca** [comment]
@god-of-hellfire root is only used if you specify it, i'm not sure your assumptions make sense.

Can you use and actual task as an example?

**Comment by bcoca** [comment]
@f-koehler your comment seems completely unrelated and the docs specify why you require `become: yes`

**Comment by god-of-hellfire** [comment]
Thanks for your replies. After further investigation and understanding what we need and want, we eventually found out how to solve this issue for us. 

First of all we changed our weird privilege escalation approach from:
* "sudo su - \<user\> -c '\<cmd\>'" to 
* "sudo -u \<user\> -i '\<cmd\>'"

But this change alone did not solve the problems for us with core modules. They still need root when they try to copy or chown or chmod things. After crawling the logs we learned that's because of the security mechanism of ansible, that forces files to be 'private' in TMP. We overcome that by changing the ansible configuration file and set the key 'allow_world_readable_tmpfiles' to 'True'. Now everything works fine for us. We are aware that the changing of the config key means to open a possible security gap but we use this only on non public systems. Besides the files will be unlinked after ansible finishes.

After all we have two experiences:
* ansible can be configured as we want it (+)
* the documentation and best practices and real-world examples are unsufficient (-)
